omnimove_external_control/src/omnimove_external_control.cpp

```diff
@@ -0,0 +1,357 @@
+#include "omnimove/omnimove_external_control.hpp"
+#include <boost/array.hpp>
+#include <boost/circular_buffer.hpp>
+#include <rclcpp/logger.hpp>
+#include <rclcpp/logging.hpp>
+#include "omnimove/external_control_message.hpp"
+
+using namespace hardware_interface;
+using namespace rclcpp_lifecycle::node_interfaces;
+using namespace boost::asio::ip;
+using namespace std;
+namespace omnimove
+{
+
+OmnimoveExternalControl::OmnimoveExternalControl()
+: hardware_interface::SystemInterface(), read_buffer_(4096)
+{
+  rclcpp::get_logger("OmnimoveExternalControl") = rclcpp::get_logger("OmnimoveExternalControl");
+}
+
+OmnimoveExternalControl::~OmnimoveExternalControl()
+{
+  if (acceptor_->is_open())
+  {
+    acceptor_->close();
+  }
+
+  if (client_socket_->is_open())
+  {
+    client_socket_->close();
+  }
+}
+
+LifecycleNodeInterface::CallbackReturn OmnimoveExternalControl::on_init(
+  const hardware_interface::HardwareInfo & info)
+{
+  if (SystemInterface::on_init(info) != LifecycleNodeInterface::CallbackReturn::SUCCESS)
+  {
+    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
+  }
+
+  agv_type_ = info.hardware_parameters.at("agv_type");
+
+  if (agv_type_ == "caterpillar")
+  {
+    position_state_.resize(5, 0);
+    velocity_state_.resize(7, 0);
+    position_commands_.resize(5, 0);
+    velocity_commands_.resize(7, 0);
+  }
+  else
+  {
+    velocity_state_.resize(3, 0);
+    velocity_commands_.resize(3, 0);
+  }
+
+  protocol_version_ = info_.hardware_parameters["protocol_version"];  // should be 1.6
+  external_control_port_ = std::stoi(info_.hardware_parameters["port"]);
+  if (
+    info.hardware_parameters.find("velocity_command_timeout_ms") != info.hardware_parameters.end())
+  {
+    vel_cmd_timeout_ms_ = std::stoi(info_.hardware_parameters["velocity_command_timeout_ms"]);
+  }
+  else
+  {
+    RCLCPP_INFO(
+      rclcpp::get_logger("OmnimoveExternalControl"), "velocity timeout not set. Using 500 ms");
+
+    vel_cmd_timeout_ms_ = 500;
+  }
+  RCLCPP_INFO(
+    rclcpp::get_logger("OmnimoveExternalControl"), "port of external control server:%d",
+    external_control_port_);
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+}
+
+hardware_interface::CallbackReturn OmnimoveExternalControl::on_configure(
+  const rclcpp_lifecycle::State & previous_state)
+{
+  return SystemInterface::on_configure(previous_state);
```

Comment:
I don't want to assume that, Since i have no further checks of my own, i prefer to call the parent function. If someone smart implements checks on the parent class, i would prefer to have them as well.